sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/container/ThroughputContainerController.java

```diff
@@ -0,0 +1,321 @@
+                                this.throughputResolveLevel = ThroughputResolveLevel.DATABASE;
+                            }
+
+                            return Mono.error(throwable);
+                        });
+                } else if (throughputResolveLevel == ThroughputResolveLevel.DATABASE) {
+                    return this.resolveThroughputByResourceId(this.targetDatabaseRid)
+                        .onErrorResume(throwable -> {
+                            if (this.isOfferNotConfiguredException(throwable)) {
+                                this.throughputResolveLevel = ThroughputResolveLevel.CONTAINER;
+                            }
+
+                            return Mono.error(throwable);
+                        });
+                }
+
+                // All the underlying throughput control groups are using target throughput,
+                // which is constant value, hence no need to resolve throughput
+                return Mono.empty();
+            })
+            .flatMap(throughputResponse -> {
+                this.updateMaxContainerThroughput(throughputResponse);
+                return Mono.empty();
+            })
+            .retryWhen(
+                // Throughput can be configured on database level or container level
+                // Retry at most 1 time so we can try on database and container both
+                RetrySpec.max(1).filter(throwable -> this.isOfferNotConfiguredException(throwable))
+            ).thenReturn(this);
+    }
+
+    private Mono<ThroughputResponse> resolveThroughputByResourceId(String resourceId) {
+        // TODO: figure out how this work for serveless account
+        // TODO: work item: https://github.com/Azure/azure-sdk-for-java/issues/18776
+        checkArgument(StringUtils.isNotEmpty(resourceId), "ResourceId can not be null or empty");
+        return this.client.queryOffers(resourceId, new CosmosQueryRequestOptions())
+            .single()
+            .flatMap(offerFeedResponse -> {
+                if (offerFeedResponse.getResults().isEmpty()) {
+                    return Mono.error(
+                        BridgeInternal.createCosmosException(NO_OFFER_EXCEPTION_STATUS_CODE, "No offers found for the resource " + resourceId));
+                }
+
+                return this.client.readOffer(offerFeedResponse.getResults().get(0).getSelfLink()).single();
+            })
+            .onErrorResume(throwable -> {
+                System.out.println(throwable.getCause());
+                return Mono.error(throwable);
+            })
+            .map(ModelBridgeInternal::createThroughputRespose);
+    }
+
+    private void setDefaultGroupController() {
+        List<ThroughputGroupControllerBase> defaultGroupControllers =
+            this.groupControllers.values().stream().filter(ThroughputGroupControllerBase::isUseByDefault).collect(Collectors.toList());
+
+        if (defaultGroupControllers.size() > 1) {
```

Comment:
What if there is no `defaultGroupController` ? Do we need to handle that case here ?

Reply:
It is okie if the customer do not want to have a default control group, no hard restriction here.